cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp

```diff
@@ -256,157 +224,108 @@ sycl::event update_centroids(sycl::queue& q,
     const auto data_ptr = values.get_data();
     const auto row_ofs_ptr = row_offsets.get_data();
     const auto col_ind_ptr = column_indices.get_data();
-    const auto counts_ptr = cluster_counts.get_data();
+    const auto counters_ptr = counters.get_data();
 
-    const auto local_size = bk::device_max_wg_size(q);
     const auto num_clusters = centroids.get_dimension(0);
 
-    const auto clean_range =
-        bk::make_multiple_nd_range_2d({ num_clusters, column_count }, { 1, 1 });
-    auto clean_event = q.submit([&](sycl::handler& cgh) {
-        cgh.depends_on(deps);
-        cgh.parallel_for(clean_range, [=](auto it) {
-            const auto cluster_id = it.get_global_id(0);
-            const auto col_id = it.get_global_id(1);
-            centroids_ptr[cluster_id * column_count + col_id] = 0;
-        });
-    });
+    ONEDAL_ASSERT_MUL_OVERFLOW(std::int64_t, num_clusters, column_count);
+    const auto centroids_elem_count = num_clusters * column_count;
+    ONEDAL_ASSERT_MUL_OVERFLOW(std::int64_t, centroids_elem_count, sizeof(Float));
+    const auto centroids_num_bytes = centroids_elem_count * sizeof(Float);
+
+    auto clean_event = q.memset(centroids_ptr, 0, centroids_num_bytes);
 
-    const auto row_block =
-        std::min<std::int32_t>(bk::device_max_wg_size(q) * 8, bk::down_pow2(row_count));
-    const auto col_block =
-        std::min<std::int32_t>(bk::device_max_wg_size(q), bk::down_pow2(column_count));
-    const auto range =
-        bk::make_multiple_nd_range_3d({ num_clusters, row_block, col_block }, { 1, 1, col_block });
+    const auto local_size = bk::device_max_sg_size(q);
+    const size_t row_count_unsigned = static_cast<size_t>(row_count);
 
+    const size_t sg_count = (row_count + local_size - 1) / local_size;
+
+    ONEDAL_ASSERT_MUL_OVERFLOW(std::int64_t, local_size, sg_count);
+    const std::int64_t global_size = sg_count * local_size;
+    auto range = bk::make_multiple_nd_range_2d({ global_size, num_clusters }, { local_size, 1 });
+
+    // Compute sum of data points belonging to each centroid
     auto centroids_sum_event = q.submit([&](sycl::handler& cgh) {
         cgh.depends_on(clean_event);
-        local_accessor_rw_t<Float> local_centroid(column_count, cgh);
-        cgh.parallel_for(range, [=](auto it) {
-            const auto cluster_id = it.get_global_id(0);
-            const auto row_shift = it.get_global_id(1);
-            const auto local_id = static_cast<std::int64_t>(it.get_local_id(2));
-            if (counts_ptr[cluster_id] == 0) {
+        cgh.depends_on(deps);
```

Comment:
you can unite it with the previous depends_on

Reply:
Done.